### ISSUE TITLE
Failed to start sparklyr backend on Databricks

### ISSUE THREAD

**Issue by kelnerrr** [sparklyr/sparklyr#2999]
---

Sparklyr's spark_connect() stopped to work on Databricks and throws an error:

> Error in value[[3L]](cond) : 
>   Failed to start sparklyr backend: java.lang.RuntimeException: Could not find a matching JAR for sparklyr among List(/databricks/spark/R/lib/sparklyr/java//sparklyr-3.1.0-preview-2.12.jar, /databricks/spark/R/lib/sparklyr/java//sparklyr-master-2.12.jar)
> 	at com.databricks.backend.daemon.driver.RDriverLocal$.com$databricks$backend$daemon$driver$RDriverLocal$$getSparklyrJarPath(RDriverLocal.scala:1228)
> 	at com.databricks.backend.daemon.driver.RDriverLocal.com$databricks$backend$daemon$driver$RDriverLocal$$launchSparklyr(RDriverLocal.scala:759)
> 	at com.databricks.backend.daemon.driver.RDriverLocal$.startSparklyr(RDriverLocal.scala:1163)
> 	at com.databricks.backend.daemon.driver.RDriverLocal.startSparklyr(RDriverLocal.scala)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.apache.spark.api.r.RBackendHandler.handleMethodCall(RBackendHandler.scala:169)
> 	at org.apache.spark.api.r.RBackendHandler.channelRead0(RBackendHandler.scala:105)
> 	at org.apache.spark.api.r.RBackendHandler.channelRead0(RBackendHandler.scala:39)
> 	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
> 	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
> 	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
> 	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
> 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	at java.lang.Thread.run(Thread.java:748)

```r
spark_version <- "3.1.1"
sc <- spark_connect(method = "databricks")
```

```
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.4 (2021-02-15)
 os       Ubuntu 18.04.5 LTS          
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  C.UTF-8                     
 ctype    C.UTF-8                     
 tz       Etc/UTC                     
 date     2021-03-30                  

─ Packages ───────────────────────────────────────────────────────────────────
 package       * version date       lib source        
 askpass         1.1     2019-01-13 [2] CRAN (R 4.0.4)
 assertthat      0.2.1   2019-03-21 [2] CRAN (R 4.0.4)
 AzureAuth       1.3.1   2021-03-30 [1] CRAN (R 4.0.4)
 AzureGraph      1.2.1   2021-02-09 [1] CRAN (R 4.0.4)
 AzureRMR      * 2.4.1   2021-02-17 [1] CRAN (R 4.0.4)
 backports       1.2.1   2020-12-09 [2] CRAN (R 4.0.4)
 base64enc       0.1-3   2015-07-28 [2] CRAN (R 4.0.4)
 broom           0.7.2   2020-10-20 [2] CRAN (R 4.0.4)
 callr           3.5.1   2020-10-13 [2] CRAN (R 4.0.4)
 catalog       * 0.1.0   2021-03-20 [1] CRAN (R 4.0.4)
 cellranger      1.1.0   2016-07-27 [2] CRAN (R 4.0.4)
 cli             2.2.0   2020-11-20 [2] CRAN (R 4.0.4)
 colorspace      2.0-0   2020-11-11 [2] CRAN (R 4.0.4)
 config          0.3     2018-03-27 [2] CRAN (R 4.0.4)
 crayon          1.3.4   2017-09-16 [2] CRAN (R 4.0.4)
 curl            4.3     2019-12-02 [2] CRAN (R 4.0.4)
 DBI             1.1.0   2019-12-15 [2] CRAN (R 4.0.4)
 dbplyr          2.0.0   2020-11-03 [2] CRAN (R 4.0.4)
 desc            1.2.0   2018-05-01 [2] CRAN (R 4.0.4)
 devtools      * 2.3.2   2020-09-18 [2] CRAN (R 4.0.4)
 digest          0.6.27  2020-10-24 [2] CRAN (R 4.0.4)
 dplyr         * 1.0.2   2020-08-18 [2] CRAN (R 4.0.4)
 ellipsis        0.3.1   2020-05-15 [2] CRAN (R 4.0.4)
 fansi           0.4.1   2020-01-08 [2] CRAN (R 4.0.4)
 fastmap         1.0.1   2019-10-08 [2] CRAN (R 4.0.4)
 forcats       * 0.5.0   2020-03-01 [2] CRAN (R 4.0.4)
 forge           0.2.0   2019-02-26 [2] CRAN (R 4.0.4)
 fs              1.5.0   2020-07-31 [2] CRAN (R 4.0.4)
 generics        0.1.0   2020-10-31 [2] CRAN (R 4.0.4)
 ggplot2       * 3.3.2   2020-06-19 [2] CRAN (R 4.0.4)
 glue            1.4.2   2020-08-27 [2] CRAN (R 4.0.4)
 gtable          0.3.0   2019-03-25 [2] CRAN (R 4.0.4)
 haven           2.3.1   2020-06-01 [2] CRAN (R 4.0.4)
 hms             0.5.3   2020-01-08 [2] CRAN (R 4.0.4)
 htmltools       0.5.0   2020-06-16 [2] CRAN (R 4.0.4)
 htmlwidgets     1.5.3   2020-12-10 [2] CRAN (R 4.0.4)
 httpuv          1.5.4   2020-06-06 [2] CRAN (R 4.0.4)
 httr            1.4.2   2020-07-20 [2] CRAN (R 4.0.4)
 hwriter         1.3.2   2014-09-10 [2] CRAN (R 4.0.4)
 hwriterPlus     1.0-3   2015-01-05 [2] CRAN (R 4.0.4)
 jose            1.0     2019-01-20 [1] CRAN (R 4.0.4)
 jsonlite        1.7.2   2020-12-09 [2] CRAN (R 4.0.4)
 later           1.1.0.1 2020-06-05 [2] CRAN (R 4.0.4)
 lifecycle       0.2.0   2020-03-06 [2] CRAN (R 4.0.4)
 lubridate     * 1.7.9.2 2020-11-13 [2] CRAN (R 4.0.4)
 magrittr        2.0.1   2020-11-17 [2] CRAN (R 4.0.4)
 memoise         1.1.0   2017-04-21 [2] CRAN (R 4.0.4)
 mime            0.9     2020-02-04 [2] CRAN (R 4.0.4)
 modelr          0.1.8   2020-05-19 [2] CRAN (R 4.0.4)
 munsell         0.5.0   2018-06-12 [2] CRAN (R 4.0.4)
 openssl         1.4.3   2020-09-18 [2] CRAN (R 4.0.4)
 pillar          1.4.7   2020-11-20 [2] CRAN (R 4.0.4)
 pkgbuild        1.1.0   2020-07-13 [2] CRAN (R 4.0.4)
 pkgconfig       2.0.3   2019-09-22 [2] CRAN (R 4.0.4)
 pkgload         1.1.0   2020-05-29 [2] CRAN (R 4.0.4)
 prettyunits     1.1.1   2020-01-24 [2] CRAN (R 4.0.4)
 processx        3.4.5   2020-11-30 [2] CRAN (R 4.0.4)
 promises        1.1.1   2020-06-09 [2] CRAN (R 4.0.4)
 ps              1.5.0   2020-12-05 [2] CRAN (R 4.0.4)
 purrr         * 0.3.4   2020-04-17 [2] CRAN (R 4.0.4)
 r2d3            0.2.3   2018-12-18 [2] CRAN (R 4.0.4)
 R6              2.5.0   2020-10-28 [2] CRAN (R 4.0.4)
 rappdirs        0.3.1   2016-03-28 [2] CRAN (R 4.0.4)
 Rcpp            1.0.5   2020-07-06 [2] CRAN (R 4.0.4)
 readr         * 1.4.0   2020-10-05 [2] CRAN (R 4.0.4)
 readxl          1.3.1   2019-03-13 [2] CRAN (R 4.0.4)
 remotes         2.2.0   2020-07-21 [2] CRAN (R 4.0.4)
 reprex          0.3.0   2019-05-16 [2] CRAN (R 4.0.4)
 rlang           0.4.9   2020-11-26 [2] CRAN (R 4.0.4)
 rprojroot       2.0.2   2020-11-15 [2] CRAN (R 4.0.4)
 Rserve          1.8-7   2021-02-17 [2] local         
 rstudioapi      0.13    2020-11-12 [2] CRAN (R 4.0.4)
 rvest           0.3.6   2020-07-25 [2] CRAN (R 4.0.4)
 scales          1.1.1   2020-05-11 [2] CRAN (R 4.0.4)
 sessioninfo     1.1.1   2018-11-05 [2] CRAN (R 4.0.4)
 shiny           1.5.0   2020-06-23 [2] CRAN (R 4.0.4)
 sparklyr      * 1.6.1   2021-03-23 [1] CRAN (R 4.0.4)
 SparkR          3.1.0   2021-03-30 [1] local         
 stringi         1.5.3   2020-09-09 [2] CRAN (R 4.0.4)
 stringr       * 1.4.0   2019-02-10 [2] CRAN (R 4.0.4)
 TeachingDemos   2.10    2016-02-12 [2] CRAN (R 4.0.4)
 testthat        3.0.0   2020-10-31 [2] CRAN (R 4.0.4)
 tibble        * 3.0.4   2020-10-12 [2] CRAN (R 4.0.4)
 tidyr         * 1.1.2   2020-08-27 [2] CRAN (R 4.0.4)
 tidyselect      1.1.0   2020-05-11 [2] CRAN (R 4.0.4)
 tidyverse     * 1.3.0   2019-11-21 [1] CRAN (R 4.0.4)
 usethis       * 2.0.0   2020-12-10 [2] CRAN (R 4.0.4)
 utf8            1.1.4   2018-05-24 [2] CRAN (R 4.0.4)
 vctrs           0.3.5   2020-11-17 [2] CRAN (R 4.0.4)
 withr           2.3.0   2020-09-22 [2] CRAN (R 4.0.4)
 xml2            1.3.2   2020-04-23 [2] CRAN (R 4.0.4)
 xtable          1.8-4   2019-04-21 [2] CRAN (R 4.0.4)
 yaml            2.2.1   2020-02-01 [2] CRAN (R 4.0.4)

[1] /databricks/spark/R/lib
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library
```

**Comment by yitao-li** [comment]
@kelnerrr `sparklyr` packages jar files built with different versions of Spark. There is no `sparklyr-3.1.0-preview-2.12.jar` yet. You can copy `sparklyr-3.0-2.12.jar` to `R/lib/sparklyr/java//sparklyr-3.1.0-preview-2.12.jar` as a workaround.

**Comment by CooperChen453** [comment]
Put another method to solve the question, as long as you don't need to use the latest features in the latest version of sparklyr :)

You can install the previous(older) version of sparklyr like this in databricks R notebook:

`devtools::install_version('sparklyr', version = '1.5.0')`

And then, it works.

**Comment by yitao-li** [comment]
@kelnerrr @CooperChen453 I think this is more of a Databricks problem: it looks like Databricks runtime has its own logic for finding the most appropriate jar files among all jar files shipped with `sparklyr` (see `RDriverLocal.scala` in the stack trace @kelnerrr included in this issue), and when it fails to find any usable `sparklyr` jar file (which really shouldn't be the case), it crashes.

I wish I could help, but given that `RDriverLocal.scala` doesn't look like something that is open-source, I think the only possible way you can resolve this is to ask for help on this from Databricks folks. Sorry.

**Comment by yitao-li** [comment]
cc @falaki @Loquats

**Comment by falaki** [comment]
@kelnerrr what version of the Databricks Runtime are you using?

**Comment by kelnerrr** [comment]
Thanks @yitao-li, @falaki, and @CooperChen453 for getting back to me.
I use Databricks 8.1 (includes Apache Spark 3.1.1, Scala 2.12).
I will check whether installing older sparklyr version help and let you know.

**Comment by kelnerrr** [comment]
I can't load sparklyr 1.5.
First it tells me that 

> Error : package or namespace load failed for ‘sparklyr’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]): Error : package or namespace load failed for ‘sparklyr’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):
>  namespace ‘rlang’ 0.4.9 is already loaded, but >= 0.4.10 is required

I tried to install newer rlang (0.4.11) but when I run `library(rlang)` it throws:

> Error in value[[3L]](cond) : Error in value[[3L]](cond) : 
>   Package ‘rlang’ version 0.4.9 cannot be unloaded:
>  Error in unloadNamespace(package) : namespace ‘rlang’ is imported by ‘usethis’, ‘htmltools’, ‘ellipsis’, ‘purrr’, ‘forge’, ‘lifecycle’, ‘pkgload’, ‘testthat’ so cannot be unloaded

All these packages seem to be loaded by Databricks after restart.

**Comment by falaki** [comment]
@kelnerrr Databricks Runtime 8.1 has sparklyr 1.5.1 pre-installed as you can find [here](https://docs.databricks.com/release-notes/runtime/8.1.html#installed-r-libraries).

Also just tested loading the package and creating a spark connection by:
```
library(sparklyr)
sc <- spark_connecdt(method = "databricks")
```

If this does not work for you, please file a ticket through. your Databricks support.

**Comment by yitao-li** [comment]
@falaki @Loquats Also a possibly related issue: someone mentioned in https://github.com/r-spark/sparklyr.flint/issues/55 a sparklyr extension is not working with Databricks connection. The same extension does work with "vanilla" Spark connections though (e.g., works on a EMR Spark cluster or similar).

My guess is the sparklyr extension tells sparklyr to fetch some Maven artifacts (i.e., jar files of some Scala library) when connecting to Spark. This seems to work well with all sorts of "vanilla" Spark connections, but when working with a Databricks cluster, the required jar files may have been downloaded to an incorrect location where the Databricks runtime is not looking for those jar files. Is there some simple way to fix this type of problem? The user in the aforementined issue ended up installing jar files manually into the Databricks cluster to make everything work, which seems non-ideal. Surely there is some more streamlined way of making it work, right?

So long story short: I think there may be something missing or not up-to-update when Javier and you all created the `spark_connect(master = "databricks")` feature, and it might be something easy to fix (hopefully) and can save people a lot of headaches.

**Comment by yitao-li** [comment]
So in essence: when I have `spark_connect(method = "databricks")`, is it possible to tell the remote Databricks cluster I'm working with to load some additional jar files (by Maven coordinate or by some dbfs file path) in addition to the usual sparklyr-*-*.jar file?

Also, if I look at https://github.com/sparklyr/sparklyr/blob/main/R/databricks_connection.R#L22 the `extensions` parameter in that function is unused. I think that may be the root of all problems with sparklyr extensions not working on Databricks connections.

**Comment by falaki** [comment]
@yitao-li I think I found the root cause. It seems the latest published package (1.6) removed the master jar, and that was our default choice when we could not find a matching JAR (based on the runtime's spark version). Databricks Runtimes 8.0 and above are running Spark 3.1.  Do we plan to compile a JAR for 3.1 or we want to continue using 3.0?

**Comment by yitao-li** [comment]
@falaki Thanks so much for the help. I'll working on creating/recovering the required jar files soon.
